swift-sdk/IterableAppIntegrationInternal.swift

```diff
@@ -9,6 +9,40 @@
 import Foundation
 import UserNotifications
 
+// Returns whether notifications are enabled
+protocol NotificationStateProviderProtocol {
+    var notificationsEnabled : Promise<Bool, Error> {get}
+    func registerForRemoteNotifications()
+}
+
+struct SystemNotificationStateProvider : NotificationStateProviderProtocol {
+    var notificationsEnabled: Promise<Bool, Error> {
+        let result = Promise<Bool, Error>()
+        
+        if #available(iOS 10.0, *) {
+            UNUserNotificationCenter.current().getNotificationSettings { settings in
+                if settings.authorizationStatus == .authorized {
+                    result.resolve(with: true)
+                } else {
+                    result.resolve(with: false)
+                }
+            }
+        } else {
+            // Fallback on earlier versions
+            if let currentSettings = UIApplication.shared.currentUserNotificationSettings, currentSettings.types != [] {
+                result.resolve(with: true)
+            } else {
+                result.resolve(with: false)
+            }
+        }
+        
+        return result
+    }
+
+    func registerForRemoteNotifications() {
+        UIApplication.shared.registerForRemoteNotifications()
```

Comment:
What do you think about wrapping this in `DispatchQueue.main.async` and remove it from `enableDeviceForCurrentUser`?